src/main/java/com/microsoft/sqlserver/jdbc/ConfigurableRetryRule.java

```diff
@@ -0,0 +1,259 @@
+    private final String ZERO = "0";
+
+    private ArrayList<Integer> waitTimes = new ArrayList<>();
+
+    /**
+     * Construct a ConfigurableRetryRule object from a String rule.
+     *
+     * @param rule
+     *        The rule used to construct the ConfigRetryRule object
+     * @throws SQLServerException
+     *         If there is a problem parsing the rule
+     */
+    public ConfigurableRetryRule(String rule) throws SQLServerException {
+        addElements(removeExtraElementsAndSplitRule(rule));
+        calculateWaitTimes();
+    }
+
+    /**
+     * Allows constructing a ConfigRetryRule object from another ConfigRetryRule object. Used when the first object has
+     * multiple errors provided. We pass in the multi-error object and create 1 new object for each error in the initial
+     * object.
+     *
+     * @param newRule
+     *        The rule used to construct the ConfigRetryRule object
+     * @param baseRule
+     *        The ConfigRetryRule object to base the new objects off of
+     */
+    public ConfigurableRetryRule(String newRule, ConfigurableRetryRule baseRule) {
+        copyFromRule(baseRule);
+        this.retryError = newRule;
+    }
+
+    private void copyFromRule(ConfigurableRetryRule baseRule) {
+        this.retryError = baseRule.retryError;
+        this.operand = baseRule.operand;
+        this.initialRetryTime = baseRule.initialRetryTime;
+        this.retryChange = baseRule.retryChange;
+        this.retryCount = baseRule.retryCount;
+        this.retryQueries = baseRule.retryQueries;
+        this.waitTimes = baseRule.waitTimes;
+    }
+
+    /**
+     * Removes extra elements in the rule (e.g. '{') and splits the rule based on ':' (colon).
+     *
+     * @param rule
+     * @return
+     */
+    private String[] removeExtraElementsAndSplitRule(String rule) {
+        if (rule.endsWith(":")) {
+            rule = rule + ZERO; // Add a zero to make below parsing easier
+        }
+
+        rule = rule.replace("{", "");
+        rule = rule.replace("}", "");
+        rule = rule.trim();
+
+        return rule.split(":"); // Split on colon
+    }
+
+    /**
+     * Checks if the value passed in is numeric. In the case where the value contains a comma, the value must be a
+     * multi-error value, e.g. 2714,2716. This must be separated, and each error checked separately.
+     *
+     * @param value
+     *        The value to be checked
+     * @throws SQLServerException
+     *         if a non-numeric value is passed in
+     */
+    private void checkParameter(String value) throws SQLServerException {
+        if (!StringUtils.isNumeric(value)) {
+            String[] arr = value.split(COMMA);
+            for (String error : arr) {
+                if (!StringUtils.isNumeric(error)) {
+                    MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_InvalidParameterFormat"));
+                    Object[] msgArgs = {error, "Not a positive number"};

```

Comment:
hardcoded error string, this should be in resource file.
Also, we already have a R_invalidParameterNumber can we re-use that?

Reply:
I didn't see that before. Replaced.